v0/destinations/engage/transform.js

```diff
@@ -0,0 +1,214 @@
+  }
+  const { publicKey, privateKey } = Config;
+  const response = defaultRequestConfig();
+  const basicAuth = Buffer.from(`${publicKey}:${privateKey}`).toString(
+    "base64"
+  );
+  response.headers = {
+    "Content-Type": "application/json",
+    Authorization: `Basic ${basicAuth}`
+  };
+  response.body.JSON = removeUndefinedAndNullValues(payload);
+  response.endpoint = endpoint;
+  response.method = method;
+  return response;
+};
+
+const identifyResponseBuilder = (message, Config) => {
+  const payload = constructPayload(
+    message,
+    mappingConfig[ConfigCategories.IDENTIFY.name]
+  );
+  const { traits } = message;
+  const uid = getUID(message);
+  // LIST FOR NOW IS IN CONFIG BUT TODO
+  const endpoint = `${ConfigCategories.IDENTIFY.endpoint.replace("uid", uid)}`;
+  const refinedPayload = refinePayload(
+    traits,
+    ConfigCategories.IDENTIFY.genericFields
+  );
+  set(payload, "meta", refinedPayload);
+  const listIds = getLists(message, Config);
+  if (listIds.length > 0) {
+    set(payload, "lists", listIds);
+  }
+  return responseBuilder(
+    payload,
+    endpoint,
+    ConfigCategories.IDENTIFY.method,
+    Config
+  );
+};
+
+const trackResponseBuilder = (message, Config) => {
+  if (!message.event) {
+    throw new CustomError("[ ENGAGE ]:: Event Name can not be empty.", 400);
+  }
+  const { properties } = message;
+  const payload = {};
+  const meta = refinePayload(properties, ConfigCategories.TRACK.genericFields);
+  const uid = getUID(message);
+  const endpoint = `${ConfigCategories.TRACK.endpoint.replace("uid", uid)}`;
+  payload.properties = meta;
+  payload.event = message.event;
+  payload.timestamp = message.originalTimestamp;
+  return responseBuilder(
+    payload,
+    endpoint,
+    ConfigCategories.TRACK.method,
+    Config
+  );
+};
+const pageResponseBuilder = (message, Config) => {
+  const uid = getUID(message);
+  const endpoint = `${ConfigCategories.PAGE.endpoint.replace("uid", uid)}`;
+
+  const { properties } = message;
+  const payload = {};
+  const meta = refinePayload(properties, ConfigCategories.PAGE.genericFields);
+  const pagePayload = constructPayload(
+    message,
+    mappingConfig[ConfigCategories.PAGE.name]
+  );
+  const mergedPayload = { ...meta, ...pagePayload };
+  payload.properties = mergedPayload;
+  payload.timestamp = message.originalTimestamp;
+  const name = getEvent(message);
+  payload.event = name;
+  return responseBuilder(
+    payload,
+    endpoint,
+    ConfigCategories.PAGE.method,
+    Config
+  );
+};
+
+const groupResponseBuilder = (message, Config) => {
+  const { groupId } = message;
+  if (!groupId) {
+    throw new CustomError("[ ENGAGE ]:: Group Id can not be empty.", 400);
+  }
+  const uid = getDestinationExternalID(message, "engageId");
+  const traits = getFieldValueFromMessage(message, "traits");
+  const operation = traits.operation ? traits.operation.toLowerCase() : "add";
```

Comment:
operation field is specific to Engage only. So that should come via integrations object.

Reply:
Not compulsory as the destinations supporting ungrouping can take help directly from this like in the case of Mautic .

Reply:
Suppose a group call is being made so every destination supporting ungrouping as well can take help from this field. Otherwise for those not supporting nothing would be affected. Also Since this value will remain same for all the destinations unlike ID's.

Reply:
For identify call we are using `getUID` we are falling back to userId right? That means the user is being created with userId, if externalID is not present. 
While sending group call, if the any random UID is sent, is a new user created and then it is put into the group? If that is the case then fine. Otherwise we can safely fallback to userId.

Reply:
So what we are currently doing is if UID is present inside externalId we are sending that inside the url `PUT /lists/{id}/subscribers/{uid}`
but if its not then we are sending the payload `POST /lists/{id}/subscribers
` and in that payload either email or phone is required by engage to make an internal lookup for the Id if they found the Id they update that user otherwise it creates a new User.

Ref:
<img width="1792" alt="Screenshot 2022-09-26 at 11 36 03 AM" src="https://user-images.githubusercontent.com/62471433/192204681-903e320e-5e71-4e4b-9357-de33d0d2358e.png">
 in this if we already know the